### ISSUE TITLE
Type check ignored on a dictionary

### ISSUE THREAD

**Issue by danmur97** [python/mypy#20791]
**Bug Report**

Type check is ignored on keys and values of a dictionary when declared in a specific way.

**To Reproduce**

[playground example](https://mypy-play.net/?gist=84b167c9a3f36ee0672ee4379a09be8b)

Mypy detects the error whenever `Util.constructor` or `Util.apply` is not used. But if both are used the calls to `f` used on the keys or values are not type checked.

**Expected Behavior**
error: Argument 1 has incompatible type "str"; expected "int"  [arg-type] at the calls of `f`

**Actual Behavior**

Mypy does not throw arg-type errors
**Your Environment**

checkout playground example

**Comment by danmur97** [comment]
Also, if I translate the apply into a var, the error raises correctly at `f` calls.
```
x: dict[str, str] = Util.apply(
        "", lambda _: {}
)
Util.constructor(
      {
          f("foo"): f("not_int")
      }
      | x
)
```

**Comment by A5rocks** [comment]
Looks like we're just not checking the dictionary keys in that position, the `reveal_type` here isn't triggering anything (it should):
```
from typing import Callable, TypeVar


_K = TypeVar("_K")
_V = TypeVar("_V")

class Foo:
    pass

class Util:
    @staticmethod
    def apply(i: str, g: Callable[[str], _K]) -> _K:
        return g(i)

    @staticmethod
    def constructor(_: dict[_K, _V]) -> Foo:
        return Foo()

def proof_of_concept(
    f:  Callable[[int], str],
) -> None:
    Util.constructor(
        {
            reveal_type(f("foo")): f("not_int")
        }
        | Util.apply(
            "", lambda _: {}
        )
    )
```
